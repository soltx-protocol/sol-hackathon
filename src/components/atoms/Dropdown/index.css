.panel {
	max-height: 170px;

	/* overflow: auto; */

	overflow-x: hidden;
	background-color: white;
}

.option {
	display: flex;
	min-width: 0;
	align-items: center;
	height: 42px;
	padding: 0 12px;
	font-style: normal;
	font-weight: 500;
	font-size: 14px;
	line-height: 20px;
	color: var(--text);
	background-color: var(--white);
	cursor: pointer;
	overflow: hidden;
	white-space: nowrap;
	text-overflow: ellipsis;
	border-radius: 4px;

	&.is-selected {
		font-weight: 500;
	}

	&:hover {
		background-color: var(--neutral-control-30);
	}

	& + .option {
		margin-top: 4px;
	}
}

.value {
	display: flex;
	justify-content: space-between;
	align-items: center;
	width: 100%;
	height: 48px;
	padding: 0 12px;

	& > div {
		font-style: normal;
		font-weight: 400;
		font-size: 16px;
		line-height: 24px;
		color: var(--text-50);
	}

	& > svg {
		width: 14px;
		height: 14px;

		& path {
			fill: var(--text-40);
		}
	}

	&.has-value > div {
		color: var(--text);
		overflow: hidden;
		white-space: nowrap;
		text-overflow: ellipsis;
	}
}
