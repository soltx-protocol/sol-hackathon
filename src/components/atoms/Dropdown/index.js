/* eslint-disable react/jsx-props-no-spreading */

import React, { useState, useEffect } from 'react';
import classnames from 'classnames';

import DropdownIcon from 'images/icon/dropdown.inline.svg';

import DropdownBasic from '../DropdownBasic';

import styles from './index.css';

const useSelect = ({ initialValue, onChange, propsValue }) => {
	const [value, setState] = useState(initialValue);

	const onChangeValue = newValue => {
		setState([newValue]);
		onChange([newValue]);
	};

	useEffect(() => {
		if (propsValue) {
			setState(propsValue);
		}
	}, propsValue);

	return [value, onChangeValue];
};

const Dropdown = ({
	initialValue = [],
	value: propsValue,
	onChange = () => {},
	placeholder,
	className,
	options = [],
	...other
}) => {
	const [value, onChangeValue] = useSelect({ initialValue, onChange, propsValue });

	const hasValue = value.length > 0 && options.some(item => item.value === value[0].value);

	return (
		<DropdownBasic
			className={className}
			valueClassName={classnames(styles.value, { [styles.hasValue]: hasValue })}
			valueComponent={() => (
				<>
					<div>{hasValue ? value[0].label : placeholder}</div>
					<DropdownIcon />
				</>
			)}
			panelClassName={styles.panel}
			panelComponent={({ closePanel }) => (
				<>
					{options.map(option => (
						<div
							key={option.value}
							className={classnames(styles.option, {
								[styles.isSelected]: value.length > 0 && option.value === value[0].value,
							})}
							onClick={() => {
								onChangeValue(option);
								closePanel();
							}}
							role="presentation"
						>
							{option.label}
						</div>
					))}
				</>
			)}
			{...other}
		/>
	);
};

export default Dropdown;
