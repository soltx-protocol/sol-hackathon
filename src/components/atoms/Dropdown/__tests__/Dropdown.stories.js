import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, object } from '@storybook/addon-knobs/react';

import Dropdown from 'components/atoms/Dropdown';

const options = [
	{ value: 'black', label: '黑色' },
	{ value: 'red', label: '紅色' },
	{ value: 'blue', label: '藍色' },
];

const stories = storiesOf('COMMON|atoms/Dropdown', module);

stories.addDecorator(withKnobs);

const selectedOptions = [{ value: 'black', label: '黑色' }];

stories.add('__interactive', () => (
	<Dropdown
		placeholder="請選擇顏色"
		options={object('list', options)}
		initialValue={object('selected value', selectedOptions)}
	/>
));

stories.add('with placeholder', () => (
	<Dropdown placeholder="請選擇顏色" options={options} initialValue={[]} />
));
