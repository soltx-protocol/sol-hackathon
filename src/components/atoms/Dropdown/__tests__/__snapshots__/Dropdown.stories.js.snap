// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <div
    class="dropdownBasic"
  >
    <div
      class="content value hasValue"
      role="button"
      tabindex="0"
    >
      <div>
        黑色
      </div>
      <test-file-stub />
    </div>
  </div>
</DocumentFragment>
`;

exports[`with placeholder 1`] = `
<DocumentFragment>
  <div
    class="dropdownBasic"
  >
    <div
      class="content value"
      role="button"
      tabindex="0"
    >
      <div>
        請選擇顏色
      </div>
      <test-file-stub />
    </div>
  </div>
</DocumentFragment>
`;
